sdks/python/apache_beam/coders/coders.py

```diff
@@ -1136,6 +1136,62 @@ def __hash__(self):
     common_urns.coders.WINDOWED_VALUE.urn, WindowedValueCoder)
 
 
+class ParamWindowedValueCoder(WindowedValueCoder):
+  """A coder used for parameterized Windowed values."""
+
+  def __init__(self, payload, components):
+    super(ParamWindowedValueCoder, self).__init__(components[0], components[1])
+    self._wrapped_value_coder = components[0]
+    self.window_coder = components[1]
+    self.payload = payload
+    self.components = components
+
+  def _create_impl(self):
+    return coder_impl.ParamWindowedValueCoderImpl(
+        self._wrapped_value_coder.get_impl(),
+        self.window_coder.get_impl(),
+        self.payload)
+
+  def is_deterministic(self):
+    return all(c.is_deterministic() for c in [self._wrapped_value_coder,
+                                              self.window_coder])
+
+  def as_cloud_object(self, coders_context=None):
```

Comment:
This is meant to support Dataflow explicitly and is similar to the proto conversion.

You'll need to add one to java in [CloudObjectTranslators](https://github.com/apache/beam/blob/a06b7a6cbad2386b7bee37d3be0d7f3f5b8bfaff/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslators.java#L244) and wire it together.

You'll need to add the encoded payload as part of the cloud object as well and I suggest you add it as a 'payload' field. Please use the same byte string -> json compatible string conversion found elsewhere in these as_cloud_object conversions.

Reply:
Thanks a lot for the detailed suggestions. How about implementing it in a following PR? We can throw an exception in this PR for now. Involve too much in a single PR seems not a good idea. What do you think?

Reply:
Seems fair to fix this at a later point in time, since this is not used in Dataflow at the moment.